server/logging/simple_logger.go

```diff
@@ -17,211 +17,199 @@ package logging
+	return l.history.String()
 }
 
-// Underlying returns the underlying logger.
-func (l *SimpleLogger) Underlying() *log.Logger {
-	return l.Logger
+func (l *StructuredLogger) Debug(format string, a ...interface{}) {
+	l.z.Debugf(format, a...)
+	l.saveToHistory(Debug, format, a...)
 }
 
-// GetLevel returns the current log level of the logger.
-func (l *SimpleLogger) GetLevel() LogLevel {
-	return l.Level
+func (l *StructuredLogger) Info(format string, a ...interface{}) {
+	l.z.Infof(format, a...)
+	l.saveToHistory(Info, format, a...)
 }
 
-func (l *SimpleLogger) saveToHistory(level string, msg string) {
-	l.History.WriteString(fmt.Sprintf("[%s] %s\n", level, msg))
+func (l *StructuredLogger) Warn(format string, a ...interface{}) {
+	l.z.Warnf(format, a...)
+	l.saveToHistory(Warn, format, a...)
 }
 
-func (l *SimpleLogger) capitalizeFirstLetter(s string) string {
-	runes := []rune(s)
-	runes[0] = unicode.ToUpper(runes[0])
-	return string(runes)
+func (l *StructuredLogger) Err(format string, a ...interface{}) {
+	l.z.Errorf(format, a...)
+	l.saveToHistory(Error, format, a...)
 }
 
-// levelToString returns the logging level as a 4 character string. DEBUG and ERROR are shortened intentionally
-// so that logs line up.
-func (l *SimpleLogger) levelToString(level LogLevel) string {
+func (l *StructuredLogger) Log(level LogLevel, format string, a ...interface{}) {
 	switch level {
 	case Debug:
-		return "DBUG"
+		l.Debug(format, a...)
 	case Info:
-		return "INFO"
+		l.Info(format, a...)
 	case Warn:
-		return "WARN"
+		l.Warn(format, a...)
 	case Error:
-		return "EROR"
+		l.Err(format, a...)
 	}
-	return "????"
 }
 
-// callSite returns the location of the caller of this function via its
-// filename and line number. skip is the number of stack frames to skip.
-func (l *SimpleLogger) callSite(skip int) (string, int) {
-	_, file, line, ok := runtime.Caller(skip)
-	if !ok {
-		return "???", 0
+func (l *StructuredLogger) SetLevel(lvl LogLevel) {
+	if l != nil {
+		l.level.SetLevel(lvl.zLevel)
 	}
+}
+
+func (l *StructuredLogger) Flush() error {
+	return l.z.Sync()
+}
 
-	// file is the full filepath but we just want the filename.
-	// NOTE: rather than calling path.Base we're using code from the stdlib
-	// logging package which I assume is optimized.
-	short := file
-	for i := len(file) - 1; i > 0; i-- {
-		if file[i] == '/' {
-			short = file[i+1:]
-			break
-		}
+func (l *StructuredLogger) saveToHistory(lvl LogLevel, format string, a ...interface{}) {
+	if !l.keepHistory {
+		return
 	}
-	return short, line
+	msg := fmt.Sprintf(format, a...)
+	l.history.WriteString(fmt.Sprintf("[%s] %s\n", lvl.shortStr, msg))
 }
+
+// NewNoopLogger creates a logger instance that discards all logs and never
+// writes them. Used for testing.
+func NewNoopLogger(t *testing.T) SimpleLogging {
+	level := zap.DebugLevel
+	return &StructuredLogger{
+		z:     zaptest.NewLogger(t, zaptest.Level(level)).Sugar(),
+		level: zap.NewAtomicLevelAt(level),
+	}
+}
```

Comment:
thinking out loud here, does it make more sense to `t.Log` these?

Reply:
I imagine that's probably what this is doing under the hood but haven't verified. This just makes it easy to conform to the same interface without creating an additional implementation. What are the benefits of using t.Log directly?

Reply:
ahh I missed that we're passing `t` into zaptest.NewLogger. Carry on.